### PR TITLE
Add find_device function for utilize Device trait

### PR DESCRIPTION
**Pull request by justapig9020**
Use "find_device" and "find_device_mut" functions which return a
implemented "Device" trait. Therefore we can reach load/store utility
with a single function call calling by the returned trait.

### CODE REVIEW

**Comment by yodalee** [comment]
Seems good.
I prefer add two properties to Memory structure in memory.rs: read-only and invalid.
If set read-only, the store operation do nothing.
If set invalid, it shows message "Read/Write at unusable address".

By this we can make UNUSABLE region an invalid memory, and CATRIDGE region as read-only memory.

**Comment by justapig9020** [comment]
I have no idea how `LR35902` handle an invalid access (included write to a read-only memory). However, in my experiment on `arm` processor, the access will caused an `bus error`. Should we categorize all invalid access as an error?

**Comment by yodalee** [comment]
We don't have to raise error when access on invalid memory. It will cause error and halt the vm. Just like what I did, use `info!` to print a message and return `Ok`.
The memory is called "unusable memory". It is suggested not to use it, but some game like Tetris DO use it for some unknown reason. We don't know what will happen if we use it. To know the actual behavior, we may have to reverse engineer the gameboy hardware.
I also leave some comment on your change, there are a few more change need to do. After that I will accept the PR.

File: src/bus.rs [hunk 1]
```diff
@@ -158,77 +158,99 @@ impl Bus {
...
+            _ => return None,
+        }
+        Some(ret)
+    }
+
+    fn load(&self, addr: u16) -> Result<u8, ()> {
+        match self.find_device(addr) {
+            Some(dev) => dev.load(addr),
+            None => match addr {
+                UNUSABLE_START ..= UNUSABLE_END => {
```
Comment (yodalee) [inline src/bus.rs]: After your change of memory permission flag, make UNUSALBE a Memory struct and mark it as invalid.
We want to put any possible region into your find_device.

File: src/bus.rs [hunk 2]
```diff
@@ -158,77 +158,99 @@ impl Bus {
         self.timer.is_interrupt = (value >> TIMER_SHIFT)  & 0x4 != 0;
     }
 
-    fn load(&self, addr: u16) -> Result<u8, ()> {
+    fn find_device(&self, addr: u16) -> Option<&dyn Device> {
+        let ret: &dyn Device;
```
Comment (yodalee) [inline src/bus.rs]: Is this ret necessary? Can we write something like this?
```rust
match addr {
    CATRIDGE_START ..= CATRIDGE_END => Some(&self.catridge),
    VRAM_START ..= VRAM_END => Some(&self.gpu),
    //...
    _ => None,
}
```

File: src/bus.rs [hunk 3]
```diff
@@ -158,77 +158,99 @@ impl Bus {
...
+
+    fn store(&mut self, addr: u16, value: u8) -> Result<(), ()> {
+        match self.find_device_mut(addr) {
+            Some(dev) => dev.store(addr, value),
+            None => match addr {
+                CATRIDGE_START ..= CATRIDGE_END => Ok(()), // catridge is read-only
+                UNUSABLE_START ..= UNUSABLE_END => {
+                    info!("Write at unusable address {:#X}", addr);
+                    Ok(())
+                }
```
Comment (yodalee) [inline src/bus.rs]: Like comment above, make cartridge read-only memory and unusable invalid memory.

File: src/bus.rs (hunk 1 above)
Comment (justapig9020) [reply src/bus.rs]: According to my test binary trying to access some UNUSABLE memory, so I marked the area as NORMAL rather than  Invalid.

Comment (justapig9020) [reply src/bus.rs]: I read your comment clearly once again, and I found out that I was misunderstood about your description of `Invalid memory`. I am going to push another commit to fix it.

**Comment by yodalee** [comment]
Yes, it is called unusable, and suggest not to use by manual. However some binary JUST use it.
I think we just use NORMAL memory, and remove the Permission flag Invalid. If accessing some memory area will cause error, we simply not creating the memory structure. The bus will return Err if it cannot find proper device.
Can you squash your change to one commit, I will merge it after that.

**Review by yodalee — COMMENTED** [review]
I think it is fine, just use it as NORMAL memory.

**Comment by justapig9020** [comment]
So, in conclusion, remove the `Invalid` flag and makes UNUSABLE memory NORMAL?

**Comment by yodalee** [comment]
Maybe we stop changing now. We can make it NORMAL memory later.
Please squash the changes to merge.

**Comment by yodalee** [comment]
I found that I can do the squash in github actually lol. It really have some new function that I have not used before.

**Comment by justapig9020** [comment]
I just found that something goes wrong in the last commit. I modify macro `err!` into `info!` without including it. The bug has fixed in my latest commit in the same branch. What should I do now to update the modification? And I apologize for didn't double check before push the commit.

**Comment by yodalee** [comment]
Just file another pull request fix the bug, thanks.

**Pull request merged**